storage/core/src/adaptor/locking.rs

```diff
@@ -0,0 +1,157 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+//! A simple adaptor to add transaction capability to a type that only implements the basic
+//! read/write operations, giving a full-featured (albeit not necessarily efficient) backend.
+
+use crate::{
+    adaptor::{Construct, CoreOps},
+    backend as traits,
+    info::{DbDesc, DbIndex},
+    Data,
+};
+use traits::{ReadOps, WriteOps};
+
+use common::sync;
+use std::collections::BTreeMap;
+
+// Read-only transaction just holds a read lock to the database
+pub struct TxRo<'tx, T>(sync::RwLockReadGuard<'tx, T>);
+
+impl<'tx, T: ReadOps> ReadOps for TxRo<'tx, T> {
+    fn get(&self, idx: DbIndex, key: &[u8]) -> crate::Result<Option<&[u8]>> {
+        self.0.get(idx, key)
+    }
+}
+
+impl<'tx, T: ReadOps> traits::TxRo for TxRo<'tx, T> {}
+
+// Tracker for database changes
+type DeltaMap = BTreeMap<Data, Option<Data>>;
+
+// RW transaction holds a write lock to the database and a list of changes performed
+pub struct TxRw<'tx, T> {
+    db: sync::RwLockWriteGuard<'tx, T>,
+    deltas: Vec<DeltaMap>,
+}
+
+impl<'tx, T> TxRw<'tx, T> {
+    fn update(&mut self, idx: DbIndex, key: Data, val: Option<Data>) -> crate::Result<()> {
+        self.deltas[idx.get()].insert(key, val);
+        Ok(())
+    }
+}
+
+impl<'tx, T: ReadOps> ReadOps for TxRw<'tx, T> {
+    fn get(&self, idx: DbIndex, key: &[u8]) -> crate::Result<Option<&[u8]>> {
+        self.deltas[idx.get()]
+            .get(key)
+            .map_or_else(|| self.db.get(idx, key), |x| Ok(x.as_deref()))
+    }
+}
+
+impl<'tx, T> WriteOps for TxRw<'tx, T> {
+    fn put(&mut self, idx: DbIndex, key: Data, val: Data) -> crate::Result<()> {
+        self.update(idx, key, Some(val))
+    }
+
+    fn del(&mut self, idx: DbIndex, key: &[u8]) -> crate::Result<()> {
+        self.update(idx, key.to_vec(), None)
+    }
+}
+
+impl<'tx, T: ReadOps + WriteOps> traits::TxRw for TxRw<'tx, T> {
+    fn commit(mut self) -> crate::Result<()> {
+        for (idx, kvmap) in self.deltas.into_iter().enumerate().map(|(i, m)| (DbIndex::new(i), m)) {
+            for (key, val) in kvmap {
+                match val {
+                    None => self.db.del(idx, &key)?,
+                    Some(val) => self.db.put(idx, key, val)?,
+                }
+            }
+        }
+        Ok(())
+    }
+}
+
+pub struct TransactionLockImpl<T> {
+    db: sync::Arc<sync::RwLock<T>>,
+    num_maps: usize,
+}
+
+impl<T> Clone for TransactionLockImpl<T> {
+    fn clone(&self) -> Self {
+        Self {
+            db: sync::Arc::clone(&self.db),
+            num_maps: self.num_maps,
+        }
+    }
+}
+
+impl<'tx, T: 'tx + ReadOps> traits::TransactionalRo<'tx> for TransactionLockImpl<T> {
```

Comment:
Generally speaking, transactions don't need to be thread-safe, and it'd be great if we could have it that way by default. Usually you start a transaction in one thread, and end it in the same thread. Even in async context, we never go across boundaries. In fact, this is a huge performance boost, since only committing transactions will require locking, while the flood of things to do locally (especially writes) to that transaction object is with no locking.

Reply:
Sure. Backends are not required to use this for the transaction capability. E.g. for LMDB backend, I plan to just use the LMDB's transaction implementation, which I suppose is way more optimised than this. We can also implement other adaptors that use a more optimistic synchronisation mechanisms. This just splits the in-memory backend implementation into two more reusable components, it does not really change how it is implemented.